### PR TITLE
Refactored perception pipeline tutorial from ros1 to ros 2

### PR DESCRIPTION
**Pull request by CihatAltiparmak**
### Description

I have a PR for refactoring perception pipeline tutorial based on ROS 2. But i have several question about my PR.

1) I used some external links for not storing big files such as video, bag file which is created by me. Is it okay to give a link to my google drive for bag file and a link to Youtube video for perception pipeline demonstration or should i change these links?

2) I used [Intel's sdf file](https://github.com/IntelRealSense/realsense-ros/blob/ros2-development/realsense2_description/meshes/d435.dae)  in the tutorial which the relevant repository has Apache 2.0 License. Does that occur any problem for this organization? Btw, [ros2/launch](https://github.com/ros2/launch/blob/rolling/LICENSE) has also Apache 2.0 License. But it is not indicated in this repository. I need a feedback for this situation.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by CihatAltiparmak** [comment]
Just for friendly pinging, @sjahr

File: doc/examples/perception_pipeline/CMakeLists.txt
Comment (sjahr) [inline doc/examples/perception_pipeline/CMakeLists.txt:2]: We try to avoid the ${PROJECT_NAME} variable and use the concrete project name to avoid errors since last year. Do you mind updating that for this tutorial?

File: doc/examples/perception_pipeline/CMakeLists.txt
Comment (sjahr) [inline doc/examples/perception_pipeline/CMakeLists.txt:10]: You can merge the install commands like this:

```
install(DIRECTORY urdf worlds meshes ...
  DESTINATION share/${PROJECT_NAME})
```
The ... need to be replaced with the missing directory names

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:19]: Why are you deleting all this, is it outdated?

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:18]: I wouldn't use your personal cloud. We could add the file to an additional repository with git lfs to store it. How big is it?

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:10]: Very nice that you recorded a video! Instead of linking a youtube video here, wouldn't a gif work too?
Additional, I think it would be nice if the video just contains the relevant video and not your entire screen. I use `kazam` to record videos and it is very straight forward to capture a window only

**Review by sjahr — COMMENTED** [review]
Nice work, I still need to test it but I left a couple of comments

**Comment by sjahr** [comment]
> I used [Intel's sdf file](https://github.com/IntelRealSense/realsense-ros/blob/ros2-development/realsense2_description/meshes/d435.dae) in the tutorial which the relevant repository has Apache 2.0 License. Does that occur any problem for this organization?

Good question :thinking: did you use the sdf or the mesh file or both?

File: doc/examples/perception_pipeline/CMakeLists.txt
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/CMakeLists.txt]: I changed ${PROJECT_NAME} with moveit2_tutorials. Do you think it's make sense?

File: doc/examples/perception_pipeline/CMakeLists.txt
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/CMakeLists.txt]: Done it

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:19]: I actually didn't delete it. I just made some modifications. You can find the relavant redirection in this link .

https://github.com/CihatAltiparmak/moveit2_tutorials/blob/e7eec45b14d360b8e8ffe2471b38f22e9f7e05cf/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L15-L18

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst]: Agree, but unfortunatelly the bag size is approximately 56MB. I will try to record this bag file in lower size. Do you have any suggestion about it? I'm not sure git lfs works here.

**Comment by CihatAltiparmak** [comment]
> Good question 🤔 did you use the sdf or the mesh file or both?

I had added Intel's license content to mesh file like below.

Screenshot from 2024-06-24 13-51-39

The most of the sdf is mine however i took [some camera measurement](https://github.com/CihatAltiparmak/moveit2_tutorials/blob/main/doc/examples/perception_pipeline/urdf/realsense_d435/macros.xacro) from them. If it is boring in your perspective, I may remove it.

File: doc/examples/perception_pipeline/CMakeLists.txt
Comment (sjahr) [reply doc/examples/perception_pipeline/CMakeLists.txt]: Thanks! The best practice hasn't made it yet into the tutorials package. But I'd like to start with that here

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:22]: This section shouldn't be removed

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:47]: Can we keep the param descriptions too? You could merge them as a comment with the `sensors_3d.yaml: ::` description

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:67]: We can drop the depth map since it is not covered in this tutorial anyways I guess

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:17]:
```suggestion
Even though you haven't completed :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` ever, you can still run this tutorial's demo. But, it is suggested to start with steps in :doc:`Getting Started </doc/tutorials/getting_started/getting_started>` for better understanding what's going on in this tutorial.
```

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:19]:
```suggestion
How to create 3D Pointcloud Data for Octomap Creation
```
Just a suggestion

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:21]: There's a typo in `depth_camera_envrionment.launch.py` and I don't understand the second sentence. Do you mind rewording it?

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:129]: Can we keep this sentence? Maybe move it to the introduction

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:45]:
```suggestion
In next step, we will use the recorded bag file to create an octomap.
```
Rewording suggestion

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:50]: Do you mind rewording this to be a bit more precise too? I am not sure if I understand it correctly.

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:69]: How much work would it be to add a depthcamera example here instead of the second Pointcloud updater (and can you briefly explain why we need two if that's required)?

**Review by sjahr — COMMENTED** [review]
I tested it and it works! I've added a bunch of comments and suggestions and once they are addressed we can finally merge this :rocket:

**Comment by sjahr** [comment]
@CihatAltiparmak Maybe this is helpful as a reference: https://github.com/moveit/moveit2_tutorials/pull/700/files

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:22]: Added to this part . https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L21-L24

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:47]: I have added to https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L90-L108 . I will also add for depth map plugin.

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst]: You're absolutely right. I reworded like this 

```
For recording bag, firstly, we can run ``depth_camera_envrionment.launch.py`` file and then record the bag using following commands.
```

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst]: I reworded it here and i added another part based on your suggestion.

https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L57-L62

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:81]: It's not necessary to use two pointcloud, These pointclouds show both two sides of table. We can turn second plugin into depth camera plugin. When i completed, i will tag you.

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:129]: Added in https://github.com/CihatAltiparmak/moveit2_tutorials/blob/bc1504d3d6c13e17c19bf506fe70358bbc83a100/doc/examples/perception_pipeline/perception_pipeline_tutorial.rst?plain=1#L57-L62

**Comment by CihatAltiparmak** [comment]
Hello @sjahr , i have added the depth image octomap updater to the perception pipeline tutorial. However i have found some bugs in depth image updater and i have a fix for it. Btw i've also sent PR to moveit_benchmark_resources.

My PR in moveit_benchmark_resources : https://github.com/moveit/moveit_benchmark_resources/pull/3
My PR to fix bug in depth image updater plugin: https://github.com/moveit/moveit2/pull/2963

After merging the above PR's, i believe we can finalize this work.

**Comment by CihatAltiparmak** [comment]
@sjahr I've applied your suggestion in moveit_benchmark_resources and here. I've added redirection to https://github.com/moveit/moveit_benchmark_resources repository. I hope this is the last dance here.

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst]: I've added. You can find some configurations here. 

https://github.com/CihatAltiparmak/moveit2_tutorials/blob/c5383b570f8e03157165f878dc67f27c6ec3df28/doc/examples/perception_pipeline/config/sensors_3d.yaml#L1-L22

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:67]: Added

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (CihatAltiparmak) [reply doc/examples/perception_pipeline/perception_pipeline_tutorial.rst]: Now it's not necessary. We are about to move bag file into moveit_benchmark_resources

File: doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py
Comment (sjahr) [inline doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py:157]:
```suggestion
```

File: doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py
Comment (sjahr) [inline doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py:172]:
```suggestion
```

File: doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py
Comment (sjahr) [inline doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py:162]:
```suggestion
```

File: doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py
Comment (sjahr) [inline doc/examples/perception_pipeline/launch/perception_pipeline_demo.launch.py:24]:
```suggestion
```

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:4]: Remove the numbers before the headings to match the style of the other tutorials
```suggestion
Introduction
```

File: doc/examples/perception_pipeline/perception_pipeline_tutorial.rst
Comment (sjahr) [inline doc/examples/perception_pipeline/perception_pipeline_tutorial.rst:2]: When building locally, I get a warning when this does not match the number of heading letters 
```suggestion
============================
```

**Review by sjahr — COMMENTED** [review]
Just some nitpicks. I am testing the tutorial one last time with the new files and then we'll merge it

**Comment by sjahr** [comment]
Oh and can you fix the broken link CI error? https://github.com/moveit/moveit2_tutorials/actions/runs/10389968570/job/28769302814?pr=906. Looks like you just need to redirect it to moveit_benchmarking_resources

**Comment by CihatAltiparmak** [comment]
> Oh and can you fix the broken link CI error? https://github.com/moveit/moveit2_tutorials/actions/runs/10389968570/job/28769302814?pr=906. Looks like you just need to redirect it to moveit_benchmarking_resources

I know, it's searching for bag_files/depth_camera_bag directory. We need to merge my moveit_benchmark_resources pr first to fix CI

**Comment by sjahr** [comment]
> > Oh and can you fix the broken link CI error? https://github.com/moveit/moveit2_tutorials/actions/runs/10389968570/job/28769302814?pr=906. Looks like you just need to redirect it to moveit_benchmarking_resources
> 
> I know, it's searching for bag_files/depth_camera_bag directory. We need to merge my moveit_benchmark_resources pr first to fix CI

@CihatAltiparmak I merged it ~1h ago. Can you run formatting and retrigger CI by pushing. Then this should be good to go

**Comment by CihatAltiparmak** [comment]
All CIes are passed. Let's roll.

**Review by sjahr — APPROVED** [review]

**Comment by sjahr** [comment]
Thank you for porting this!

**Comment by 130s** [comment]
@sjahr I'm glad a long-standing ticket https://github.com/moveit/moveit2_tutorials/issues/25 is finally closed by this PR.
Then, more than a year ago I made basically the same effort https://github.com/moveit/moveit2_tutorials/pull/700. You also commented you'd start reviewing https://github.com/moveit/moveit2_tutorials/pull/700#issuecomment-1645650394. I admit I hadn't kept asking for review recently though, I'm honestly a bit sad. It'd be great in the future if open PRs aren't discarded without any accolade like in this way.

**Comment by sjahr** [comment]
@130s you're absolutely right and I am very sorry that this happened due to my inattentiveness. As you pointed out I even committed to reviewing your PR last year and then never did it :disappointed: This should not happen and you're right to be frustrated about my (in)action.
I think we should re-open your PR as this merged version is a simplified version of the ROS 1 tutorial that doesn't include the Detecting and Adding Object as Collision Object section that you ported as well and use your work to enhance and improve the existing version.

**Comment by CihatAltiparmak** [comment]
I want to write as well. Firstly really apologies for this action @130s, You are absolutely right. I already saw your PR before i didn't open this PR. As a newbie of MoveIt2, i strongly needed to learn perception pipeline application in Moveit and i had limited time. I took a look at MoveIt1's tutorial and it didn't be helpful for me as newbie. When i see your PR has no active works, i decided to continue to my work. I have added depth camera octomap updater and i found 2 bugs in MoveIt2. 

But i just want to empathize the maintainer as well. When i open the review page of my PR, i saw a lot of complex works for maintainer side and a lot of PR's to review.

If you still want to help us, my PR can have some lack of information. @sjahr 's suggestions are also good.

Best for you

**Comment by 130s** [comment]
@sjahr @CihatAltiparmak You guys made my day. Thanks, thanks. I'll see what I can pick up from #700. Another good thing is now we have a new reviewer for perception pipeline :)

**Pull request merged**
